server/src/main/java/org/elasticsearch/index/seqno/RetentionLeaseBackgroundSyncAction.java

```diff
@@ -0,0 +1,185 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.seqno;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.apache.lucene.store.AlreadyClosedException;
+import org.elasticsearch.ExceptionsHelper;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.replication.ReplicationRequest;
+import org.elasticsearch.action.support.replication.ReplicationResponse;
+import org.elasticsearch.action.support.replication.TransportReplicationAction;
+import org.elasticsearch.cluster.action.shard.ShardStateAction;
+import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.IndexShardClosedException;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Replication action responsible for background syncing retention leases to replicas. This action is deliberately a replication action so
+ * that if a replica misses a background retention lease sync then that shard will not be marked as stale. We have some tolerance for a
+ * shard copy missing renewals of retention leases since the background sync interval is much smaller than the expected lifetime of
+ * retention leases.
+ */
+public class RetentionLeaseBackgroundSyncAction extends TransportReplicationAction<
```

Comment:
I am not sure if it's worth to avoid having two SyncAction instances. How about adding a new background sync flag to the Request. Then makes `newReplicasProxy` depend on that new flag. If this is not worth, how about sharing Request/Response between RetentionLeaseBackgroundSyncAction and RetentionLeaseSyncAction? But I am total fine with your current choice since the duplicate is pretty small though.

Reply:
Yeah I thought about this option instead of having two sync actions. The reason that I settled on two syncs is I really don't want to have to think about an action that is sometimes a write action and sometimes a replication action. It feels kind of hacky to override `newReplicasProxy` on write action to turn around and sometimes return a replication action `ReplicasProxy`. I am happy to share the request/responses though.

Reply:
@dnhatn Actually I would prefer to not share the request/response classes too since the request/response classes for the write action subclass replicated write request versus replication request; while a replicated write request is a replication request, I would rather not leak the write hierarchy here either. Can you review as-is?

Reply:
I am fine with your choice :).